docs/layout.rst

```diff
@@ -0,0 +1,246 @@
+
+
+AoS versus SoA
+--------------
+
+Tensors of same size can be placed together.
+
+For example, this places two 1D tensors of size ``3`` (array of structure, AoS):
+
+.. code-block:: python
+
+    ti.root.dense(ti.i, 3).place(x, y)
+
+Their memory layout:
+
+.. code-block::
+
+    #  address low ............. address high
+    #  x[0]   y[0] | x[1]  y[1] | x[2]   y[2]
+
+In contrast, this places two tensor placed separately (structure of array, SoA):
+
+.. code-block:: python
+
+    ti.root.dense(ti.i, 3).place(x)
+    ti.root.dense(ti.i, 3).place(y)
+
+Now, their memory layout:
+
+.. code-block::
+
+    #  address low ............. address high
+    #  x[0]  x[1]   x[2] | y[0]   y[1]   y[2]
+
+
+Normally, you don't have to worry about the performance nuances between different layouts, and should just define the simplest layout as a start.
+However, locality sometimes have a significant impact on the performance, especially when the tensor is huge.
+
+**To improve spatial locality of memory accesses (i.e. cache hit rate / cacheline utilization), it's sometimes helpful to minimize the address differences of elements that are accessed together.**
+Take a simple 1D wave equation solver for example:
+
+.. code-block:: python
+
+    N = 200000
+    pos = ti.var(ti.f32)
+    vel = ti.var(ti.f32)
+    ti.root.dense(ti.i, N).place(pos)
+    ti.root.dense(ti.i, N).place(vel)
+
+    @ti.kernel
+    def step():
+        pos[i] += vel[i] * dt
+        vel[i] += -k * pos[i] * dt
+
+
+Here, we placed ``pos`` and ``vel`` seperately. So the distance in address space between ``pos[i]`` and ``vel[i]`` is ``200000``. This will result in a poor spatial locality and lots of cache-misses, which damages the performance.
+A better placement is to place them together:
+
+.. code-block:: python
+
+    ti.root.dense(ti.i, N).place(pos, vel)
+
+Then ``vel[i]`` is placed right next to ``pos[i]``, this can increases the cache-hit rate and therefore increases performance.
+
+
+Flat layouts versus blocked layouts 
+-------------------------
+
+By default, when allocating a ``ti.var`` , it follows the most naive data layout
+
+.. code-block:: python
+
+  val = ti.var(ti.f32, shape=(32, 64, 128))
+  # C++ equivalent: float val[32][64][128]
+
+However, oftentimes this data layout is suboptimal for computer graphics tasks.
+For example, ``val[i, j, k]`` and ``val[i + 1, j, k]`` are very far away (``32 KB``) from each other, and leads to poor access locality under certain computation tasks. Specifically, in tasks such as texture trilinear interpolation, the two elements are not even within the same ``4KB`` pages, creating a huge cache/TLB pressure.
+
+A better layout might be
+
+.. code-block:: python
+
+  val = ti.var(ti.f32)
+  ti.root.dense(ti.ijk, (8, 16, 32)).dense(ti.ijk, (4, 4, 4)).place(val)
+
+This organizes ``val`` in ``4x4x4`` blocks, so that with high probability ``val[i, j, k]`` and its neighbours are close to each other (i.e., in the same cacheline or memory page).
```

Comment:
```suggestion
This organizes ``val`` in ``4x4x4`` blocks, so that with high probability ``val[i, j, k]`` and its neighbours are close to each other (i.e., in the same memoy page or even cacheline).
```

Reply:
Refuse, typo `memoy` -> `memory`.

Reply:
My point is that memory should come before cacheline